pyomo/repn/plugins/nl_writer.py

```diff
@@ -0,0 +1,2124 @@
+            if not linear:
+                continue
+            ostream.write(
+                f'G{obj_idx} {len(linear)}{row_comments[obj_idx + n_cons]}\n')
+            for _id in sorted(linear.keys(), key=column_order.__getitem__):
+                ostream.write(
+                    f'{column_order[_id]} {linear[_id]!r}\n'
+                )
+
+        # Generate the return information
+        info = NLWriterInfo(
+            variables, constraints, objectives, sorted(amplfunc_libraries),
+            row_labels, col_labels)
+        timer.toc("Wrote NL stream", level=logging.DEBUG)
+        timer.toc("Generated NL representation", delta=False)
+        return info
+
+    def _categorize_vars(self, comp_list, linear_by_comp):
+        """Categorize compiled expression vars into linear and nonlinear
+
+        This routine takes an iterable of compiled component expression
+        infos and returns the sets of variables appearing lineary and
+        nonlinearly in those components.
+
+        This routine has a number of side effects:
+
+          - the ``linear_by_comp`` dict is updated to contain the set of
+            nonzeros for each component in the ``comp_list``
+
+          - the expr_info (the second element in each tuple in
+            ``comp_list``) is "compiled": the ``linear`` attribute is
+            converted from a list of coef, var_id terms (potentially with
+            duplicate entries) into a dict that maps var id to
+            coefficients
+
+        Returns
+        -------
+        all_linear_vars: set
+            set of all vars that only appear linearly in the compiled
+            component expression infos
+
+        all_nonlinear_vars: set
+            set of all vars that appear nonlinearly in the compiled
+            component expression infos
+
+        nnz_by_var: dict
+            Count of the number of components that each var appears in.
+
+        """
+        all_linear_vars = set()
+        all_nonlinear_vars = set()
+        nnz_by_var = {}
+
+        for comp_info in comp_list:
+            expr_info = comp_info[1]
+            # Note: mult will be 1 here: it is either cleared by
+            # finalizeResult, or this is a named expression, in which
+            # case the mult was reset within handle_named_expression_node
+            #assert expr_info.mult == 1
+            #
+            # Process the linear portion of this component
+            if expr_info.linear:
+                if expr_info.linear.__class__ is list:
+                    linear = {}
+                    for v, c in expr_info.linear:
+                        if v in linear:
+                            linear[v] += c
+                        else:
+                            linear[v] = c
+                    expr_info.linear = linear
+                linear_vars = set(expr_info.linear)
+                all_linear_vars.update(linear_vars)
+            # else:
```

Comment:
Still necessary?

Reply:
Yes.  This is actually documentation as to why the (seemingly necessary symbol definition is not actually necessary).  Updating comment to clarify.